src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export(typeof(IPasteTrackingService)), Shared]
+    [Export(typeof(PasteTrackingService))]
+    internal class PasteTrackingService : IPasteTrackingService
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly ILightBulbBroker2 _lightBulbBroker;
+        private readonly ISuggestedActionCategorySet _refactoringsCategorySet;
+
+        [ImportingConstructor]
+        public PasteTrackingService(IThreadingContext threadingContext, ILightBulbBroker2 lightBulbBroker, ISuggestedActionCategoryRegistryService categoryRegistryService)
+        {
+            _threadingContext = threadingContext;
+            _lightBulbBroker = lightBulbBroker;
+            _refactoringsCategorySet = categoryRegistryService.AllRefactorings;
+        }
+
+        public bool TryGetPastedTextSpan(Document document, out TextSpan textSpan)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                textSpan = default;
+                return false;
+            }
+
+            // `PropertiesCollection` is thread-safe
+            return textBuffer.Properties.TryGetProperty(this, out textSpan);
+        }
+
+        internal void RegisterPastedTextSpan(ITextView textView, ITextBuffer textBuffer, TextSpan textSpan)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            textBuffer.Changed += RemovePastedTestSpan;
+            textBuffer.Properties.AddProperty(this, textSpan);
+
+            // Try to start the light bulb session now so that it displayed to the user sooner.
+            TryShowLightBulb(textView, textSpan);
+
+            return;
+
+            void RemovePastedTestSpan(object sender, TextContentChangedEventArgs e)
+            {
+                textBuffer.Changed -= RemovePastedTestSpan;
+                textBuffer.Properties.RemoveProperty(this);
+            }
+        }
+
+        private bool TryGetTextBuffer(Document document, out ITextBuffer textBuffer)
+        {
+            if (document == null ||
+                !document.TryGetText(out var text))
+            {
+                textBuffer = null;
+                return false;
+            }
+
+            textBuffer = text.Container.TryGetTextBuffer();
+            return textBuffer != null;
+        }
+
+        private void TryShowLightBulb(ITextView textView, TextSpan textSpan)
```

Comment:
you're doing a lot of UI specific work in these functions.  we should have thread affinity checks in the functions in this type.

Reply:
This method is only invoked by `RegisterPastedTextSpan` and that method uses `Contract.ThrowIfFalse(_threadingContext.HasMainThread);`  as a guard. Is there anything else in this class I need to watch out for?

Reply:
> This method is only invoked by RegisterPastedTextSpan and that method uses Contract.ThrowIfFalse(_threadingContext.HasMainThread)

That's the case *today*.  What happens if a BG method adds a call here in teh future?  The method itself should be annotated so it would break early if someone ever made that mistake :)